### PR TITLE
Rsdk 1155 nick module edit

### PR DESCRIPTION
**Pull request by nicksanford**
Depends on: https://github.com/viamrobotics/api/pull/464

1. Add `WriteRTP` to gostream stream interface
2. Change mod manager & module to attempt to create optional GRPC peer connections.
3. Change mod manager to use new SharedConn wrapper which wraps both a `ReconfigurableClientConn` and an optional `*webrtc.PeerConnection` & provides a thread safe interface to add & remove track subscriptions by resource.
4. Update API from https://github.com/viamrobotics/api/compare/v0.1.225...v0.1.227
5. Update goutils from https://github.com/viamrobotics/goutils/compare/v0.1.67...v0.1.71

### CODE REVIEW

File: module/module.go [hunk 1]
```diff
@@ -141,12 +143,21 @@ func NewHandlerMapFromProto(ctx context.Context, pMap *pb.HandlerMap, conn rpc.C
 	return hMap, errs
 }
 
+type peerResourceState struct {
+	// subscription *camera.StreamSubscription
```
Comment (nicksanford) [inline module/module.go:147]: I left comments where this will hook into to the new camera apis.

File: module/wrtc_helper.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	"go.viam.com/rdk/logging"
+)
+
+// ConfigureForRenegotiation sets up PeerConnection callbacks for updating local descriptions and
+// sending offers when a negotiation is needed (e.g: adding a video track). As well as listening for
+// offers/answers to update remote descriptions (e.g: when the peer adds a video track).
+//
+// If successful, a Go channel is returned. The Go channel will close when the negotiation
+// DataChannel is open and available for renegotiation.
+func ConfigureForRenegotiation(peerConn *webrtc.PeerConnection, logger logging.Logger) (<-chan struct{}, error) {
```
Comment (dgottlieb) [inline module/wrtc_helper.go:24]: I've put up [this PR](https://github.com/viamrobotics/goutils/pull/273) such that we can avoid adding this file.

File: gostream/stream.go [hunk 3]
```diff
@@ -31,6 +33,7 @@ type Stream interface {
 
 	// Start starts processing frames.
 	Start()
+	WriteRTP(pkt *rtp.Packet) error
```
Comment (dgottlieb) [inline gostream/stream.go:35]: Was this a thing that we were concerned was going to change the camera API?

File: webrtchack/webrtchack.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
+// TODO: This needs to also be called for remotes
+package webrtchack
```
Comment (dgottlieb) [inline webrtchack/webrtchack.go:3]: I'd like to avoid introducing a top-level package. Can this live in gostream? Or camera? Or the slightly misnomered rdk/grpc package (webrtc.go)? without introducing a cycle.

File: webrtchack/webrtchack.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+		logger:                   logger,
+	}
+	if peerConn != nil {
+		peerConn.OnTrack(func(tr *webrtc.TrackRemote, r *webrtc.RTPReceiver) {
+			name, err := resource.NewFromString(tr.StreamID())
+			if err != nil {
+				logger.Errorf("%p OnTrack called with StreamID: %s which is not able to be parsed into a resource name", sc, tr.StreamID())
+				return
+			}
+			// TODO: Maybe we don't want to hold this lock for the entire duration of the onTrackCB
```
Comment (dgottlieb) [inline webrtchack/webrtchack.go:48]: I'm assuming the `peerConnMu` is protecting access to this map? I agree with the concern and think we should just release after the map read.

File: webrtchack/webrtchack.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+func (sc *SharedConn) RemoveOnTrackSub(name resource.Name) {
+	sc.peerConnMu.Lock()
+	defer sc.peerConnMu.Unlock()
+	delete(sc.resourceOnTrackCBs, name)
+}
+
+// TODO: See if we can make this HasPeerConn() bool
+func (sc *SharedConn) PeerConn() *webrtc.PeerConnection {
+	sc.peerConnMu.Lock()
```
Comment (dgottlieb) [inline webrtchack/webrtchack.go:79]: What's the consequence of not holding this lock?

File: webrtchack/webrtchack.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	defer sc.peerConnMu.Unlock()
+	return sc.peerConn
+}
+
+func (sc *SharedConn) Close() error {
+	var err error
+	sc.peerConnMu.Lock()
```
Comment (dgottlieb) [inline webrtchack/webrtchack.go:87]: Same "what's the consequence" question.

File: module/module.go [hunk 8]
```diff
@@ -554,6 +665,100 @@ func (m *Module) OperationManager() *operation.Manager {
...
+	// err = vcss.SubscribeRTP(ctx, sub, func(pkts []*rtp.Packet) error {
+	// 	for _, pkt := range pkts {
+	// 		if err := tlsRTP.WriteRTP(pkt); err != nil {
+	// 			m.logger.Warn(err.Error())
+	// 		}
+	// 	}
+	// 	return nil
+	// })
+
+	if err != nil {
```
Comment (dgottlieb) [inline module/module.go]: I suppose this doesn't matter, but I assume this was intended to be commented out?

File: gostream/stream.go (hunk 3 above)
Comment (nicksanford) [reply gostream/stream.go:35]: no:

This is:
https://github.com/viamrobotics/rdk/pull/3723/files#diff-4eab9b93978c09650defc23c4274a4c77d87cfdea7775d86dcf8c838ec207a01R85-R98

File: webrtchack/webrtchack.go (hunk 4 above)
Comment (nicksanford) [reply webrtchack/webrtchack.go:3]: Yeah I intentionally named it something silly so we could have a convo on where it should live. 

I'll try to put it in: rdk/grpc  (as that is what it is most related to) and report back if it creates a cycle & we need to reconsider

File: module/wrtc_helper.go (hunk 2 above)
Comment (dgottlieb) [reply module/wrtc_helper.go:24]: New tag is published: https://github.com/viamrobotics/goutils/releases/tag/v0.1.71

**Comment by nicksanford** [comment]
@dgottlieb to all the `what is the consequence of the SharedConn mutex not being held` questions.

The answer is that I don't know.

TL;DR:

Do you want me to either:
1. Try to reduce the amount of locking
2. Try to replace `rdkgrpc.ReconfigurableClientConn` with this struct

Details:
I can re test removing all locks, other those that are protectecting the maps in this struct if you would like me to go down the first path.

I created the SharedConn b/c 2 concurrent camera GRPC clients were overwriting each other's OnTrack callback registration resulting in only one camera stream being dropped on the floor.

That indicates to me that it is dangerous to have unsynchronized access to peer connection objects as the pion API is not thread-safe.

The API I would like GRPC clients have is:
1. A GRPC client gets passed a connection
2. A connection can optionally be WebRTC enabled and allows clients to discover whether it is or not
3. The connection's API is thread-safe
4. If a connection's state has changed (such as the peer connection going away & coming back due to a module crash or reconfiguration) subscriptions the client has made against the connection can be recovered / cleaned up

I haven't yet implemented 4.

To your point about why we need the locks: `lock all access to either the grpc or peer conenctions` was my attempt to uphold that api.

This is somewhat confounded by the fact that the GRPC connection has its own mutex protecting it which leads to the risk of deadlocks.

The options I see are:
1. Live with the unsafety & try to lock as little as possble
2. Replace `rdkgrpc.ReconfigurableClientConn` with one that can optionally hold the webrtc peer connection so that we can think about the two connections as one atomic unit.

**Review by dgottlieb — APPROVED** [review]
lgtm % offline conversation
